In [1]:
# 경고(Warning)가 출력되지 않도록 설정.
import warnings
warnings.filterwarnings(action='ignore')

In [2]:
import pandas as pd
import numpy as np

In [3]:
data = pd.read_csv('data/data-v01.csv')
data.shape

(150000, 11)

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150000 entries, 0 to 149999
Data columns (total 11 columns):
 #   Column                                Non-Null Count   Dtype  
---  ------                                --------------   -----  
 0   seriousdlqin2yrs                      150000 non-null  int64  
 1   revolvingutilizationofunsecuredlines  150000 non-null  float64
 2   age                                   150000 non-null  int64  
 3   numberoftime30-59dayspastduenotworse  150000 non-null  int64  
 4   debtratio                             150000 non-null  float64
 5   monthlyincome                         150000 non-null  float64
 6   numberofopencreditlinesandloans       150000 non-null  int64  
 7   numberoftimes90dayslate               150000 non-null  int64  
 8   numberrealestateloansorlines          150000 non-null  int64  
 9   numberoftime60-89dayspastduenotworse  150000 non-null  int64  
 10  numberofdependents                    150000 non-null  float64
dtype

In [5]:
data.head()

seriousdlqin2yrs  revolvingutilizationofunsecuredlines  age  \
0                 1                              0.766127   45   
1                 0                              0.957151   40   
2                 0                              0.658180   38   
3                 0                              0.233810   30   
4                 0                              0.907239   49   

   numberoftime30-59dayspastduenotworse  debtratio  monthlyincome  \
0                                     2   0.802982         9120.0   
1                                     0   0.121876         2600.0   
2                                     1   0.085113         3042.0   
3                                     0   0.036050         3300.0   
4                                     1   0.024926        12645.0   

   numberofopencreditlinesandloans  numberoftimes90dayslate  \
0                               13                        0   
1                                4                        0   
2                                2                        1   
3                                5                        0   
4                                7                        0   

   numberrealestateloansorlines  numberoftime60-89dayspastduenotworse  \
0                             6                                     0   
1                             0                                     0   
2                             0                                     0   
3                             0                                     0   
4                             1                                     0   

   numberofdependents  
0                 2.0  
1                 1.0  
2                 0.0  
3                 0.0  
4                 0.0

In [6]:
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import GradientBoostingClassifier, RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from xgboost import XGBClassifier

from sklearn.pipeline import make_pipeline

from sklearn.model_selection import train_test_split, GridSearchCV, RandomizedSearchCV
from sklearn.preprocessing import StandardScaler, MinMaxScaler
from sklearn.metrics import roc_auc_score, accuracy_score, recall_score, precision_score

import pickle

In [7]:
# X, y 분리
y = data.seriousdlqin2yrs
X = data.drop('seriousdlqin2yrs', axis='columns')

In [8]:
X_train, X_test, y_train, y_test = train_test_split(X, y, stratify=y, random_state=0)

In [9]:
y_train.shape,y_test.shape

((112500,), (37500,))

In [13]:
# y의 label별 비율
np.unique(y_train, return_counts=True)[1]/y_train.size

array([0.93315556, 0.06684444])

## Feature Scaler생성

In [14]:
scaler = StandardScaler()
# scaler = MinMaxScaler()

## Base-line 모델 정의

In [16]:
# knn. losgistic은 scaler 적용
knn = make_pipeline(scaler, KNeighborsClassifier())
lr = make_pipeline(scaler, LogisticRegression(max_iter=2000, random_state=0))
rf = RandomForestClassifier(random_state=0)
grb = GradientBoostingClassifier(random_state=0)
xgb = XGBClassifier(random_state=0)

In [17]:
knn.fit(X_train, y_train)
lr.fit(X_train, y_train)
xgb.fit(X_train, y_train)
grb.fit(X_train, y_train)
rf.fit(X_train, y_train)

[11:16:26] WARNING: ..\src\learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.


RandomForestClassifier(random_state=0)

In [18]:
base_line = [knn, lr, xgb, grb, rf]
model_names = ['KNN', 'LogisticRegression', 'XGBoost', 'GradientBoosting', 'RandomForest']

In [19]:
for model, name in zip(base_line, model_names):

    pred_train = model.predict(X_train)
    pred_test = model.predict(X_test)
    
    pred_train_proba = model.predict_proba(X_train)
    pred_test_proba = model.predict_proba(X_test)
    
    acc_train = np.round(accuracy_score(y_train,pred_train),3)
    acc_test = np.round(accuracy_score(y_test, pred_test), 3)
    
    auc_train = np.round(roc_auc_score(y_train, pred_train_proba[:, 1]), 3)
    auc_test = np.round(roc_auc_score(y_test, pred_test_proba[:, 1]), 3)
    
    print(f'{name}')
    print(f'train정확도:{acc_train}, Test정확도:{acc_test}\t train AUC:{auc_train}, Test AUC:{auc_test}')
    print('='*50)

KNN
train정확도:0.943, Test정확도:0.932	 train AUC:0.952, Test AUC:0.691
LogisticRegression
train정확도:0.935, Test정확도:0.934	 train AUC:0.804, Test AUC:0.8
XGBoost
train정확도:0.949, Test정확도:0.935	 train AUC:0.916, Test AUC:0.856
GradientBoosting
train정확도:0.939, Test정확도:0.936	 train AUC:0.869, Test AUC:0.863
RandomForest
train정확도:0.999, Test정확도:0.935	 train AUC:1.0, Test AUC:0.835


# GridSearchCV를 이용한 하이퍼파라미터 튜닝

### XGBoost

In [36]:
param = {
    'learning_rate':[0.01,0.1,0.5,1],
    'n_estimators':[100,200,300,400,500],
    'max_depth':range(1,6),
    'subsample':[0.6,0.7,0.8,0.9,1],
}

In [37]:
rs_xgb = RandomizedSearchCV(XGBClassifier(random_state=0), 
                            param, 
                            n_iter=60, 
                            scoring='roc_auc',
                            cv=5,
                            n_jobs=-1 )

In [38]:
rs_xgb.fit(X_train, y_train)

[09:30:29] WARNING: D:\bld\xgboost-split_1637426510059\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.


RandomizedSearchCV(estimator=XGBClassifier(base_score=None, booster=None,
                                           colsample_bylevel=None,
                                           colsample_bynode=None,
                                           colsample_bytree=None,
                                           enable_categorical=False, gamma=None,
                                           gpu_id=None, importance_type=None,
                                           interaction_constraints=None,
                                           learning_rate=None,
                                           max_delta_step=None, max_depth=None,
                                           min_child_weight=None, missing=nan,
                                           monotone_constraints=None...
                                           num_parallel_tree=None,
                                           predictor=None, random_state=0,
                                           reg_alpha=None, 

In [40]:
rs_xgb.best_params_

{'subsample': 0.7, 'n_estimators': 400, 'max_depth': 2, 'learning_rate': 0.1}

In [41]:
rs_df = pd.DataFrame(rs_xgb.cv_results_)
rs_df.sort_values('rank_test_score').head()

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
40      31.016063      0.124241         0.053126        0.007654   
48      34.638435      0.222048         0.056251        0.007657   
52      31.075674      0.226477         0.053121        0.007653   
18      31.512466      0.226502         0.059377        0.006251   
26      41.981191      0.485211         0.043751        0.006251   

   param_subsample param_n_estimators param_max_depth param_learning_rate  \
40             0.7                400               2                 0.1   
48             0.9                300               3                 0.1   
52             0.9                200               4                 0.1   
18             0.9                400               2                 0.1   
26             0.9                500               2                 0.1   

                                               params  split0_test_score  \
40  {'subsample': 0.7, 'n_estimators': 400, 'max_d...           0.874950   
48  {'subsample': 0.9, 'n_estimators': 300, 'max_d...           0.874765   
52  {'subsample': 0.9, 'n_estimators': 200, 'max_d...           0.875298   
18  {'subsample': 0.9, 'n_estimators': 400, 'max_d...           0.874672   
26  {'subsample': 0.9, 'n_estimators': 500, 'max_d...           0.874923   

    split1_test_score  split2_test_score  split3_test_score  \
40           0.865705           0.863905           0.861238   
48           0.866375           0.862712           0.861983   
52           0.865890           0.863654           0.860988   
18           0.865861           0.863543           0.861622   
26           0.865680           0.863518           0.861644   

    split4_test_score  mean_test_score  std_test_score  rank_test_score  
40           0.862181         0.865596        0.004920                1  
48           0.862104         0.865588        0.004863                2  
52           0.861891         0.865544        0.005157                3  
18           0.861418         0.865423        0.004894                4  
26           0.861127         0.865378        0.005033                5

In [43]:
# 상세하게 찾기
param = {
    "subsample":[0.6,0.7,0.8,0.9,1], 
    "max_depth":[2,3,4]
}
gs_xgb = GridSearchCV(XGBClassifier(n_estimators=400, learning_rate=0.1, random_state=0), 
                      param, 
                      scoring='roc_auc', 
                      cv=5,
                      n_jobs=-1 )

gs_xgb.fit(X_train, y_train)

[09:58:40] WARNING: D:\bld\xgboost-split_1637426510059\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.


GridSearchCV(estimator=XGBClassifier(base_score=None, booster=None,
                                     colsample_bylevel=None,
                                     colsample_bynode=None,
                                     colsample_bytree=None,
                                     enable_categorical=False, gamma=None,
                                     gpu_id=None, importance_type=None,
                                     interaction_constraints=None,
                                     learning_rate=0.1, max_delta_step=None,
                                     max_depth=None, min_child_weight=None,
                                     missing=nan, monotone_constraints=None,
                                     n_estimators=400, n_jobs=None,
                                     num_parallel_tree=None, predictor=None,
                                     random_state=0, reg_alpha=None,
                                     reg_lambda=None, scale_pos_weight=None,
                

In [44]:
gs_xgb.best_params_

{'max_depth': 2, 'subsample': 0.6}

In [45]:
gs_xgb.best_score_

0.8657477556440014

In [46]:
gs_df = pd.DataFrame(gs_xgb.cv_results_)
gs_df.sort_values('rank_test_score').head()

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0      31.058689      0.205362         0.058918        0.008004   
9      42.705779      0.630386         0.064576        0.004157   
1      33.984716      3.338388         0.062907        0.014241   
2      38.034212      0.305476         0.076201        0.020585   
7      47.057385      0.136800         0.064977        0.012758   

  param_max_depth param_subsample                              params  \
0               2             0.6  {'max_depth': 2, 'subsample': 0.6}   
9               3               1    {'max_depth': 3, 'subsample': 1}   
1               2             0.7  {'max_depth': 2, 'subsample': 0.7}   
2               2             0.8  {'max_depth': 2, 'subsample': 0.8}   
7               3             0.8  {'max_depth': 3, 'subsample': 0.8}   

   split0_test_score  split1_test_score  split2_test_score  split3_test_score  \
0           0.875436           0.865609           0.864516           0.861334   
9           0.874870           0.866800           0.862795           0.861687   
1           0.874950           0.865705           0.863905           0.861238   
2           0.875271           0.866079           0.863023           0.861621   
7           0.875787           0.865944           0.862766           0.860993   

   split4_test_score  mean_test_score  std_test_score  rank_test_score  
0           0.861844         0.865748        0.005102                1  
9           0.861934         0.865617        0.004980                2  
1           0.862181         0.865596        0.004920                3  
2           0.861661         0.865531        0.005133                4  
7           0.861947         0.865487        0.005412                5

In [47]:
best_model_xgb = gs_xgb.best_estimator_

In [21]:
# 모델저장
import os
save_dir = 'saved_model'
os.makedirs(save_dir, exist_ok=True) #디렉토리 만들기(없으면 만들고 있으면 안 만든다.)


xgb_file_path = os.path.join(save_dir, 'xgb_best.pkl')
best_model_xgb.save_model(xgb_file_path) #xgb객체.save_model(경로) => 모델을 파일로 저장.

In [22]:
# Load Model
saved_xgb = XGBClassifier()
saved_xgb.load_model(xgb_file_path)

### GradientBoosting

In [51]:
param_gb = {
    'learning_rate':[0.001, 0.01,0.1,0.5,1,10],
    'n_estimators':[100,200,300,400,500],
    'max_depth':range(1,6),
    'subsample':[0.6,0.7,0.8,0.9,1],
}
rs_gb = RandomizedSearchCV(GradientBoostingClassifier(random_state=0), 
                           param_distributions=param_gb, 
                           n_iter=60, 
                           cv=5, 
                           scoring='roc_auc', 
                           n_jobs=-1)

In [52]:
rs_gb.fit(X_train, y_train)

RandomizedSearchCV(cv=5, estimator=GradientBoostingClassifier(random_state=0),
                   n_iter=60, n_jobs=-1,
                   param_distributions={'learning_rate': [0.001, 0.01, 0.1, 0.5,
                                                          1, 10],
                                        'max_depth': range(1, 6),
                                        'n_estimators': [100, 200, 300, 400,
                                                         500],
                                        'subsample': [0.6, 0.7, 0.8, 0.9, 1]},
                   scoring='roc_auc')

In [53]:
rs_gb.best_params_

{'subsample': 0.7, 'n_estimators': 400, 'max_depth': 3, 'learning_rate': 0.1}

In [54]:
rs_gb.best_score_

0.8649254493129028

In [55]:
rs_df2 = pd.DataFrame(rs_gb.cv_results_)
rs_df2.sort_values('rank_test_score').head()

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
4       87.439586      0.393772         0.218750        0.000003   
9       80.648487      1.104507         0.209383        0.007656   
30      95.200368      0.479226         0.218750        0.009882   
22     151.157891      0.614488         0.481239        0.006265   
3       76.538395      0.624363         0.187497        0.000005   

   param_subsample param_n_estimators param_max_depth param_learning_rate  \
4              0.7                400               3                 0.1   
9              0.7                500               2                 0.1   
30               1                500               2                 0.1   
22             0.6                500               5                0.01   
3                1                400               2                 0.1   

                                               params  split0_test_score  \
4   {'subsample': 0.7, 'n_estimators': 400, 'max_d...           0.874596   
9   {'subsample': 0.7, 'n_estimators': 500, 'max_d...           0.874537   
30  {'subsample': 1, 'n_estimators': 500, 'max_dep...           0.874283   
22  {'subsample': 0.6, 'n_estimators': 500, 'max_d...           0.874642   
3   {'subsample': 1, 'n_estimators': 400, 'max_dep...           0.874333   

    split1_test_score  split2_test_score  split3_test_score  \
4            0.865694           0.863766           0.860195   
9            0.864986           0.863223           0.860554   
30           0.865131           0.863204           0.859948   
22           0.865064           0.862633           0.859184   
3            0.864844           0.862957           0.859751   

    split4_test_score  mean_test_score  std_test_score  rank_test_score  
4            0.860377         0.864925        0.005263                1  
9            0.860563         0.864772        0.005163                2  
30           0.860498         0.864613        0.005186                3  
22           0.861454         0.864595        0.005370                4  
3            0.860694         0.864516        0.005220                5

In [61]:
param={
    "subsample":[0.6, 0.7, 0.8], 
    "n_estimators":[300,400,500],
    "max_depth":[2,3,4]
}
gs_gb = GridSearchCV(GradientBoostingClassifier(learning_rate=0.1, random_state=0),
                    param_grid=param, 
                    scoring='roc_auc',
                    cv=5,
                    n_jobs=-1)

In [62]:
gs_gb.fit(X_train, y_train)

GridSearchCV(cv=5, estimator=GradientBoostingClassifier(random_state=0),
             n_jobs=-1,
             param_grid={'max_depth': [2, 3, 4],
                         'n_estimators': [300, 400, 500],
                         'subsample': [0.6, 0.7, 0.8]},
             scoring='roc_auc')

In [63]:
gs_gb.best_params_

{'max_depth': 3, 'n_estimators': 400, 'subsample': 0.8}

In [64]:
gs_gb.best_score_

0.8651901976338401

In [65]:
gs_df2 = pd.DataFrame(gs_gb.cv_results_)
gs_df2.sort_values('rank_test_score').head()

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
14      94.270927      0.613004         0.218747        0.000011   
17     118.147506      0.748996         0.265629        0.009872   
11      76.431602      1.126797         0.177856        0.008967   
16     105.908822      0.339065         0.262501        0.006252   
13      85.046958      0.418646         0.212501        0.007652   

   param_max_depth param_n_estimators param_subsample  \
14               3                400             0.8   
17               3                500             0.8   
11               3                300             0.8   
16               3                500             0.7   
13               3                400             0.7   

                                               params  split0_test_score  \
14  {'max_depth': 3, 'n_estimators': 400, 'subsamp...           0.874113   
17  {'max_depth': 3, 'n_estimators': 500, 'subsamp...           0.874049   
11  {'max_depth': 3, 'n_estimators': 300, 'subsamp...           0.874041   
16  {'max_depth': 3, 'n_estimators': 500, 'subsamp...           0.874683   
13  {'max_depth': 3, 'n_estimators': 400, 'subsamp...           0.874596   

    split1_test_score  split2_test_score  split3_test_score  \
14           0.866089           0.863399           0.860763   
17           0.865394           0.863408           0.860701   
11           0.866068           0.863220           0.860193   
16           0.865589           0.863736           0.860127   
13           0.865694           0.863766           0.860195   

    split4_test_score  mean_test_score  std_test_score  rank_test_score  
14           0.861586         0.865190        0.004821                1  
17           0.861606         0.865032        0.004787                2  
11           0.861567         0.865018        0.004918                3  
16           0.860571         0.864941        0.005273                4  
13           0.860377         0.864925        0.005263                5

In [74]:
best_model_gb = gs_gb.best_estimator_

In [95]:
# 모델 저장
gb_file_path = os.path.join(save_dir, 'gradient_boosting_best.pkl')
with open(gb_file_path, 'wb') as fw:
    pickle.dump(best_model_gb, fw) 

In [24]:
gb_file_path = os.path.join(save_dir, 'gradient_boosting_best.pkl')
with open(gb_file_path, 'rb') as fr:
    saved_gb = pickle.load(fr) 

### RandomForest

In [56]:
param_rf = {
    'n_estimators':[100,200,300,400,500],
    'max_depth':range(1,5),
    'max_features':range(5,11)
}
rs_rf = RandomizedSearchCV(RandomForestClassifier(random_state=0), 
                                 param_distributions=param_rf, 
                                 n_iter=60, 
                                 cv=5, 
                                 scoring='roc_auc', 
                                 n_jobs=-1)

In [57]:
rs_rf.fit(X_train, y_train)

RandomizedSearchCV(cv=5, estimator=RandomForestClassifier(random_state=0),
                   n_iter=60, n_jobs=-1,
                   param_distributions={'max_depth': range(1, 5),
                                        'max_features': range(5, 11),
                                        'n_estimators': [100, 200, 300, 400,
                                                         500]},
                   scoring='roc_auc')

In [58]:
rs_rf.best_params_

{'n_estimators': 500, 'max_features': 5, 'max_depth': 4}

In [59]:
rs_rf.best_score_

0.8574212595493437

In [60]:
rs_df3 = pd.DataFrame(rs_rf.cv_results_)
rs_df3.sort_values('rank_test_score').head()

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
7       73.531847      2.457526         1.368750        0.021194   
24      40.256247      0.420611         0.824998        0.006249   
9       80.320407      1.818594         1.550004        0.121598   
1       30.761284      1.485742         0.549998        0.011696   
26      61.568321      0.602533         1.090620        0.006253   

   param_n_estimators param_max_features param_max_depth  \
7                 500                  5               4   
24                300                  5               4   
9                 500                  6               4   
1                 200                  6               4   
26                400                  6               4   

                                               params  split0_test_score  \
7   {'n_estimators': 500, 'max_features': 5, 'max_...           0.867495   
24  {'n_estimators': 300, 'max_features': 5, 'max_...           0.867615   
9   {'n_estimators': 500, 'max_features': 6, 'max_...           0.866967   
1   {'n_estimators': 200, 'max_features': 6, 'max_...           0.867609   
26  {'n_estimators': 400, 'max_features': 6, 'max_...           0.866947   

    split1_test_score  split2_test_score  split3_test_score  \
7            0.857701           0.857847           0.851686   
24           0.857822           0.857946           0.851409   
9            0.856554           0.857327           0.851474   
1            0.856124           0.856800           0.851451   
26           0.856386           0.857085           0.851530   

    split4_test_score  mean_test_score  std_test_score  rank_test_score  
7            0.852378         0.857421        0.005658                1  
24           0.852292         0.857417        0.005776                2  
9            0.852273         0.856919        0.005522                3  
1            0.852574         0.856912        0.005722                4  
26           0.852166         0.856823        0.005522                5

In [79]:
param = {
    "n_estimators":[400,500,600,700], 
    "max_features":[3,4,5,6],
    "max_depth":[3,4,5,6]
}

gs_rf = GridSearchCV(RandomForestClassifier(random_state=0),
                    param_grid=param, 
                    scoring='roc_auc',
                    cv=5,
                    n_jobs=-1)

In [80]:
gs_rf.fit(X_train, y_train)

GridSearchCV(cv=5, estimator=RandomForestClassifier(random_state=0), n_jobs=-1,
             param_grid={'max_depth': [3, 4, 5, 6],
                         'max_features': [3, 4, 5, 6],
                         'n_estimators': [400, 500, 600, 700, 700]},
             scoring='roc_auc')

In [81]:
gs_rf.best_params_

{'max_depth': 6, 'max_features': 3, 'n_estimators': 400}

In [82]:
gs_rf.best_score_

0.861018496990928

In [83]:
gs_df3 = pd.DataFrame(gs_rf.cv_results_)
gs_df3.sort_values('rank_test_score').head()

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
60      51.003557      0.361240         1.300001        0.011693   
61      64.879531      0.439918         1.649999        0.030620   
62      78.449902      0.316428         1.962504        0.015931   
63      91.269520      0.806867         2.283340        0.007059   
64      90.540809      0.366533         2.275007        0.015934   

   param_max_depth param_max_features param_n_estimators  \
60               6                  3                400   
61               6                  3                500   
62               6                  3                600   
63               6                  3                700   
64               6                  3                700   

                                               params  split0_test_score  \
60  {'max_depth': 6, 'max_features': 3, 'n_estimat...           0.871344   
61  {'max_depth': 6, 'max_features': 3, 'n_estimat...           0.871227   
62  {'max_depth': 6, 'max_features': 3, 'n_estimat...           0.871321   
63  {'max_depth': 6, 'max_features': 3, 'n_estimat...           0.871279   
64  {'max_depth': 6, 'max_features': 3, 'n_estimat...           0.871279   

    split1_test_score  split2_test_score  split3_test_score  \
60           0.861713           0.861344           0.854574   
61           0.861697           0.861102           0.854534   
62           0.861691           0.861041           0.854550   
63           0.861667           0.861067           0.854563   
64           0.861667           0.861067           0.854563   

    split4_test_score  mean_test_score  std_test_score  rank_test_score  
60           0.856118         0.861018        0.005878                1  
61           0.856227         0.860958        0.005825                2  
62           0.856165         0.860954        0.005865                3  
63           0.856156         0.860946        0.005848                4  
64           0.856156         0.860946        0.005848                4

In [84]:
best_model_rf = gs_rf.best_estimator_

In [85]:
rf_file_path = os.path.join(save_dir, "random_forest_best.pkl")
with open(rf_file_path, 'wb') as fo:
    pickle.dump(best_model_rf, fo)

In [25]:
rf_file_path = os.path.join(save_dir, "random_forest_best.pkl")
with open(rf_file_path, 'rb') as fo:
    saved_rf = pickle.load(fo)


## VotingClassifier
- best model들 사용

In [106]:
from sklearn.ensemble import VotingClassifier
estimators = [('xgb', saved_xgb), ('gradient boost', saved_gb), ('random forest', saved_rf)]
voting_clf = VotingClassifier(estimators=estimators, voting='soft', n_jobs=-1)

In [107]:
voting_clf.fit(X_train, y_train)

VotingClassifier(estimators=[('xgb',
                              XGBClassifier(base_score=0.5, booster='gbtree',
                                            colsample_bylevel=1,
                                            colsample_bynode=1,
                                            colsample_bytree=1,
                                            enable_categorical=False, gamma=0,
                                            gpu_id=-1, importance_type=None,
                                            interaction_constraints='',
                                            learning_rate=0.1, max_delta_step=0,
                                            max_depth=2, min_child_weight=1,
                                            missing=nan,
                                            monotone_constraints='()',
                                            n_estimators=40...
                                            predictor='auto', random_state=0,
                                     

In [109]:
pred_proba  = voting_clf.predict_proba(X_test)
auc_score = roc_auc_score(y_test, pred_proba[:, 1])
auc_score

0.8618094979187302

- xgboost : 0.8657477556440014
- grandient boosting : 0.8651901976338401
- RandomForest: 0.861018496990928
- voting: 0.8618094979187302

## Test Set 으로 검증

In [26]:
def test(estimator, X, y):
    pred_prob = estimator.predict_proba(X)
    return roc_auc_score(y, pred_prob[:, 1])

In [27]:
# XGB
test(saved_xgb, X_test, y_test)

0.870387089730717

In [28]:
# Gradient Boosting
test(saved_gb, X_test, y_test)

0.8734470316904399

### Test set 최종 검증결과
- xgboost : 0.870387089730717
- grandient boosting : 0.8734470316904399